File: initool.lua
#!/usr/bin/env lua

local function trim(s)
  return (s:gsub("^[ \t\r\n]+", ""):gsub("[ \t\r\n]+$", ""))
end

local function to_lower(s)
  return string.lower(s)
end

local function starts_with(s, c)
  return s ~= "" and s:sub(1, 1) == c
end

local function is_section_header(s)
  return #s >= 3 and s:sub(1, 1) == "[" and s:sub(-1) == "]"
end

local function unquote(s)
  if #s >= 2 and s:sub(1, 1) == '"' and s:sub(-1) == '"' then
    return s:sub(2, -2)
  end
  return s
end

local function quote_if_needed(s)
  if s:find("[ =]") then
    return '"' .. s .. '"'
  end
  return s
end

local IniFile = {}
IniFile.__index = IniFile

function IniFile.new(path)
  local self = setmetatable({}, IniFile)
  self.path = path
  self.lines = {}
  self.section_lines = {}
  self.key_lines = {}
  self.data = {}
  self:load()
  return self
end

function IniFile:get(section, key)
  local sec = to_lower(section)
  local k = to_lower(key)
  if self.data[sec] and self.data[sec][k] ~= nil then
    return self.data[sec][k]
  end
  error("Error: key not found")
end

function IniFile:set(section, key, value)
  local sec_lower = to_lower(section)
  local key_lower = to_lower(key)
  local val = value

  if self.section_lines[sec_lower] == nil then
    if #self.lines > 0 and trim(self.lines[#self.lines]) ~= "" then
      table.insert(self.lines, "")
    end
    table.insert(self.lines, "[" .. section .. "]")
    table.insert(self.lines, key .. " = " .. quote_if_needed(val))
    self:write()
    return
  end

  local section_line = self.section_lines[sec_lower]
  if self.key_lines[sec_lower] == nil then
    self.key_lines[sec_lower] = {}
  end
  local keys = self.key_lines[sec_lower]

  if keys[key_lower] ~= nil then
    local line_no = keys[key_lower]
    local line = self.lines[line_no]
    local pos = line:find("=", 1, true)
    local lhs
    if pos then
      lhs = trim(line:sub(1, pos - 1))
    else
      lhs = trim(line)
    end
    self.lines[line_no] = lhs .. " = " .. quote_if_needed(val)
    self:write()
    return
  end

  local insert_at = section_line + 1
  while insert_at <= #self.lines do
    local t = trim(self.lines[insert_at])
    if starts_with(t, "[") then
      if insert_at > section_line + 1 and trim(self.lines[insert_at - 1]) == "" then
        insert_at = insert_at - 1
      end
      break
    end
    insert_at = insert_at + 1
  end

  table.insert(self.lines, insert_at, key .. " = " .. quote_if_needed(val))
  self:write()
end

function IniFile:delete(section, key)
  local sec_lower = to_lower(section)
  local key_lower = to_lower(key)

  if self.key_lines[sec_lower] == nil then
    error("Error: section not found")
  end

  local keys = self.key_lines[sec_lower]
  local line_no = keys[key_lower]
  if line_no == nil then
    error("Error: key not found")
  end

  table.remove(self.lines, line_no)

  if self.data[sec_lower] then
    self.data[sec_lower][key_lower] = nil
  end
  keys[key_lower] = nil

  for _, sec_keys in pairs(self.key_lines) do
    for k, lineno in pairs(sec_keys) do
      if lineno > line_no then
        sec_keys[k] = lineno - 1
      end
    end
  end

  for sec, lineno in pairs(self.section_lines) do
    if lineno > line_no then
      self.section_lines[sec] = lineno - 1
    end
  end

  self:write()
end

function IniFile:load()
  local f = io.open(self.path, "r")
  if not f then
    error("Error: cannot open file " .. self.path)
  end

  self.lines = {}
  for line in f:lines() do
    table.insert(self.lines, line)
  end
  f:close()

  self.section_lines = {}
  self.key_lines = {}
  self.data = {}

  local current_section = ""
  for lineno, line in ipairs(self.lines) do
    local t = trim(line)
    if t ~= "" and not starts_with(t, ";") then
      if is_section_header(t) then
        current_section = to_lower(trim(t:sub(2, -2)))
        self.section_lines[current_section] = lineno
      elseif current_section ~= "" then
        local pos = t:find("=", 1, true)
        if pos then
          local k = to_lower(trim(t:sub(1, pos - 1)))
          local v = unquote(trim(t:sub(pos + 1)))
          if self.data[current_section] == nil then
            self.data[current_section] = {}
          end
          self.data[current_section][k] = v
          if self.key_lines[current_section] == nil then
            self.key_lines[current_section] = {}
          end
          self.key_lines[current_section][k] = lineno
        end
      end
    end
  end
end

function IniFile:write()
  local f = io.open(self.path, "w")
  if not f then
    error("Error: cannot write file " .. self.path)
  end

  for _, line in ipairs(self.lines) do
    f:write(line)
    f:write("\n")
  end
  f:close()
end

local function usage(argv0)
  io.stderr:write(
    "\nUsage:\n"
      .. "  " .. argv0 .. " -g, --get <file> <section> <key>\n"
      .. "  " .. argv0 .. " -s, --set <file> <section> <key> <value>\n"
      .. "  " .. argv0 .. " -d, --del <file> <section> <key>\n\n"
  )
end

local function main(argv)
  local ok, err = pcall(function()
    if #argv < 1 then
      usage(arg and arg[-1] or "initool.lua")
      os.exit(1)
    end

    local command = argv[1]
    local argv0 = arg and arg[-1] or "initool.lua"

    if command == "--get" or command == "-g" then
      if #argv ~= 4 then
        io.stderr:write("Usage: " .. argv0 .. " --get <file> <section> <key>\n")
        os.exit(1)
      end
      local ini = IniFile.new(argv[2])
      io.write(ini:get(argv[3], argv[4]))
      return
    end

    if command == "--set" or command == "-s" then
      if #argv ~= 5 then
        io.stderr:write("Usage: " .. argv0 .. " --set <file> <section> <key> <value>\n")
        os.exit(1)
      end
      local ini = IniFile.new(argv[2])
      ini:set(argv[3], argv[4], argv[5])
      io.write("Updated [" .. argv[3] .. "] " .. argv[4] .. " = " .. argv[5] .. "\n")
      return
    end

    if command == "--del" or command == "-d" then
      if #argv ~= 4 then
        io.stderr:write("Usage: " .. argv0 .. " --del <file> <section> <key>\n")
        os.exit(1)
      end
      local ini = IniFile.new(argv[2])
      ini:delete(argv[3], argv[4])
      io.write("Deleted [" .. argv[3] .. "] " .. argv[4] .. "\n")
      return
    end

    io.stderr:write("Unknown command: " .. command .. "\n")
    os.exit(1)
  end)

  if not ok then
    io.stderr:write(tostring(err) .. "\n")
    return 1
  end
  return 0
end

-- Build argv like python: argv[1] is command, etc.
local argv = {}
for i = 1, #arg do
  argv[i] = arg[i]
end

os.exit(main(argv))

File: initool.py
#!/usr/bin/env python3

from __future__ import annotations

import sys
from dataclasses import dataclass, field
from pathlib import Path


def trim(s: str) -> str:
    return s.strip(" \t\r\n")


def to_lower(s: str) -> str:
    return s.lower()


def starts_with(s: str, c: str) -> bool:
    return bool(s) and s[0] == c


def is_section_header(s: str) -> bool:
    return len(s) >= 3 and s[0] == "[" and s[-1] == "]"


def unquote(s: str) -> str:
    if len(s) >= 2 and s[0] == '"' and s[-1] == '"':
        return s[1:-1]
    return s


def quote_if_needed(s: str) -> str:
    if any(ch in s for ch in (" ", "=")):
        return f'"{s}"'
    return s


@dataclass
class IniFile:
    path: Path
    lines: list[str] = field(default_factory=list)
    section_lines: dict[str, int] = field(default_factory=dict)
    key_lines: dict[str, dict[str, int]] = field(default_factory=dict)
    data: dict[str, dict[str, str]] = field(default_factory=dict)

    def __post_init__(self) -> None:
        self.load()

    def get(self, section: str, key: str) -> str:
        sec = to_lower(section)
        k = to_lower(key)
        if sec in self.data and k in self.data[sec]:
            return self.data[sec][k]
        raise RuntimeError("Error: key not found")

    def set(self, section: str, key: str, value: str) -> None:
        sec_lower = to_lower(section)
        key_lower = to_lower(key)
        val = value

        if sec_lower not in self.section_lines:
            if self.lines and trim(self.lines[-1]) != "":
                self.lines.append("")
            self.lines.append(f"[{section}]")
            self.lines.append(f"{key} = {quote_if_needed(val)}")
            self.write()
            return

        section_line = self.section_lines[sec_lower]
        keys = self.key_lines.setdefault(sec_lower, {})

        if key_lower in keys:
            line_no = keys[key_lower]
            line = self.lines[line_no]
            pos = line.find("=")
            lhs = trim(line[:pos]) if pos != -1 else trim(line)
            self.lines[line_no] = f"{lhs} = {quote_if_needed(val)}"
            self.write()
            return

        insert_at = section_line + 1
        while insert_at < len(self.lines):
            t = trim(self.lines[insert_at])
            if starts_with(t, "["):
                if insert_at > section_line + 1 and trim(self.lines[insert_at - 1]) == "":
                    insert_at -= 1
                break
            insert_at += 1

        self.lines.insert(insert_at, f"{key} = {quote_if_needed(val)}")
        self.write()

    def delete(self, section: str, key: str) -> None:
        sec_lower = to_lower(section)
        key_lower = to_lower(key)

        if sec_lower not in self.key_lines:
            raise RuntimeError("Error: section not found")

        keys = self.key_lines[sec_lower]
        if key_lower not in keys:
            raise RuntimeError("Error: key not found")

        line_no = keys[key_lower]
        del self.lines[line_no]

        if sec_lower in self.data:
            self.data[sec_lower].pop(key_lower, None)
        keys.pop(key_lower, None)

        for _sec, _keys in self.key_lines.items():
            for _k, _lineno in list(_keys.items()):
                if _lineno > line_no:
                    _keys[_k] = _lineno - 1

        for _sec, _lineno in list(self.section_lines.items()):
            if _lineno > line_no:
                self.section_lines[_sec] = _lineno - 1

        self.write()

    def load(self) -> None:
        try:
            with self.path.open("r", encoding="utf-8") as f:
                raw_lines = f.read().splitlines()
        except OSError:
            raise RuntimeError(f"Error: cannot open file {self.path}") from None

        self.lines = list(raw_lines)
        self.section_lines.clear()
        self.key_lines.clear()
        self.data.clear()

        current_section = ""
        for lineno, line in enumerate(self.lines):
            t = trim(line)
            if t == "" or starts_with(t, ";"):
                continue

            if is_section_header(t):
                current_section = to_lower(trim(t[1:-1]))
                self.section_lines[current_section] = lineno
                continue

            if current_section:
                pos = t.find("=")
                if pos != -1:
                    k = to_lower(trim(t[:pos]))
                    v = unquote(trim(t[pos + 1 :]))
                    self.data.setdefault(current_section, {})[k] = v
                    self.key_lines.setdefault(current_section, {})[k] = lineno

    def write(self) -> None:
        try:
            with self.path.open("w", encoding="utf-8", newline="\n") as f:
                for line in self.lines:
                    f.write(line)
                    f.write("\n")
        except OSError:
            raise RuntimeError(f"Error: cannot write file {self.path}") from None


def usage(argv0: str) -> None:
    sys.stderr.write(
        "\nUsage:\n"
        f"  {argv0} -g, --get <file> <section> <key>\n"
        f"  {argv0} -s, --set <file> <section> <key> <value>\n"
        f"  {argv0} -d, --del <file> <section> <key>\n\n"
    )


def main(argv: list[str]) -> int:
    try:
        if len(argv) < 2:
            usage(argv[0])
            return 1

        command = argv[1]
        if command in ("--get", "-g"):
            if len(argv) != 5:
                sys.stderr.write(f"Usage: {argv[0]} --get <file> <section> <key>\n")
                return 1
            ini = IniFile(Path(argv[2]))
            sys.stdout.write(ini.get(argv[3], argv[4]))
            return 0

        if command in ("--set", "-s"):
            if len(argv) != 6:
                sys.stderr.write(f"Usage: {argv[0]} --set <file> <section> <key> <value>\n")
                return 1
            ini = IniFile(Path(argv[2]))
            ini.set(argv[3], argv[4], argv[5])
            sys.stdout.write(f"Updated [{argv[3]}] {argv[4]} = {argv[5]}\n")
            return 0

        if command in ("--del", "-d"):
            if len(argv) != 5:
                sys.stderr.write(f"Usage: {argv[0]} --del <file> <section> <key>\n")
                return 1
            ini = IniFile(Path(argv[2]))
            ini.delete(argv[3], argv[4])
            sys.stdout.write(f"Deleted [{argv[3]}] {argv[4]}\n")
            return 0

        sys.stderr.write(f"Unknown command: {command}\n")
        return 1
    except Exception as e:
        sys.stderr.write(str(e) + "\n")
        return 1


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))

